### PR TITLE
chore: generate PATCH-friendly bindings

### PR DESCRIPTION
**Pull request by rb-determined-ai**
For classes in bindings.py, avoid defaulting to None via default parameter values.

Instead, explicitly detect which optional parameters are set, convert them to None in the `__init__()` body, then keep track of which parameters were explicitly set.

Then in `to_json()`, only emit parameters which were explicitly set.

This strategy should play nicely with the FieldMask behavior that grpc-gateway supports.

### CODE REVIEW

File: bindings/generate_bindings.py [hunk 1]
```diff
@@ -298,37 +306,58 @@ def gen_def(self) -> Code:
...
+            if v.type.need_parse():
+                parsed = v.type.load(f'obj["{k}"]')
+                parsed = parsed + f' if obj.get("{k}", None) is not None else None'
             else:
                 parsed = f'obj.get("{k}", None)'
-            out.append(f"""            {k}={parsed},""")
-        out += ["        )"]
+            out += [f'        if "{k}" in obj:']
+            out += [f'            kwargs["{k}"] = obj["{k}"]']
+        out += ["        return cls(**kwargs)"]
```
Comment (NicholasBlaskey) [inline bindings/generate_bindings.py:339]: the generated code is really clean!

**Review by NicholasBlaskey — APPROVED** [review]
Nice, checked with the API side and it works correctly with field masks.

**Review by hamidzr — COMMENTED** [review]
great work!  A couple of points and questions I ran into

- this seems like a usecase we want to support but it doesn't look like we are updating what's set.
```
    cf = bindings.v1ColumnFilter()
    print(cf.to_json(omit_unset=True))
    cf.name = "name"
    assert cf.to_json(omit_unset=True) != {}
```

- ordered kwargs :+1

- why split out options args
```
+        if "agentUserGroup" in obj:
  +            kwargs["agentUserGroup"] = v1AgentUserGroup.from_json(obj["agentUserGroup"]) if obj["agentUserGroup"] is not None else None
  +        return cls(**kwargs)
```

- track set or unset in a set?

```
  +        if not omit_unset or self._checkpoint_is_set:
  +            out["checkpoint"] = None if self.checkpoint is None else self.checkpoint.to_json(omit_unset)
  +        if not omit_unset or self._training_is_set:
  +            out["training"] = None if self.training is None else self.training.to_json(omit_unset)
  +        if not omit_unset or self._validation_is_set:
  +            out["validation"] = None if self.validation is None else self.validation.to_json(omit_unset)
```

- why create an attribute for each field to track set status. generates a lot of lines. these don't get set outside of the constructor and the check is a type check.
    - could we check when to_json is called? track in a set if needed?
```
* +        self._totalCheckpointSize_is_set = not isinstance(totalCheckpointSize, Unset)
```

- omit_unset type could probably be Optional[bool] otherwise these are flagged as invalid:

```
experiment.to_json(None)
experiment.to_json(omit_unset=None)
```

which might be the intended use case :thinking_face: would it be easier to use if 
None would get defaulted to False?

**Comment by rb-determined-ai** [comment]
> could we check when to_json is called? track in a set if needed?

You mean, if we let values actually have the value of `_unset`?  I would say, that would cause a _lot_ of other code to have to be unset-aware, which would be a serious pain.  Being able to access `thing.field` and get the default value if nothing was set is important to simplify the consuming code.

You could do that with `@property` getters and setters of course, but that would add many, many more lines.

And for this pattern:

```python
    cf = bindings.v1ColumnFilter()
    print(cf.to_json(omit_unset=True))
    cf.name = "name"
    assert cf.to_json(omit_unset=True) != {}
```

I think this is an antipattern.  I think that the swagger-generated classes are unergonomic and generally useless except for serializing and deserializing communications from the master.  After that, things should generally be converted into python SDK objects and those should be used for object-oriented designs.

That is to say, mutating a bindings object after creation seems like there isn't a good reason to do it in the first place.

I'll agree that this PR introduces a pretty big footgun though, and if you think it should be corrected I can think of two mechanisms:

1. Set defaults at the class level and use `"attr" in vars(self)` to decide if it is actually set or not.  This is a bit weird, but it would work I think.

    ```python
    class Thing:
        attr: Optional[Int] = None
        def __init__(self, attr: Union[Unset, None, Int] = _unset):
            if not isinstance(attr, Unset):
                self.attr = attr
        
        def to_json(self):
            out = {}
            if "attr" in vars(self):
                out["attr"] = self.attr
    ``` 

2. Use `@property` for getters (and possibly for setters, or just disable setters, which would be my preference).  This would create much more generated code.

> omit_unset type could probably be Optional[bool] otherwise these are flagged as invalid:

Wait... why would `to_json(None)` be valid?

**Comment by hamidzr** [comment]
> You mean, if we let values actually have the value of _unset? 

I had missed the goal of hiding `Unset` from users. 👍🏼 

> That is to say, mutating a bindings object after creation seems like there isn't a good reason to do it in the first place.

We talked about this offline being able to manipulate these objects internally could be useful but I don't have much recent context on how we use these in our codebase to say if we want to leave this door open or to close it.

If we assume we want to keep these instances mutable then how we deal with Unset type this might need a more changes. From here on out I assume we don't want to modify these objects and don't plan to.

> I'll agree that this PR introduces a pretty big footgun though, and if you think it should be corrected I can think of two mechanisms:

Up to you. I'd lean toward lower import speed first and then readability of the internals of the generated code.

> Wait... why would to_json(None) be valid?

I agree directly setting it to None would be a bit odd but if we normally default it to None for the user, having the option
of passing a value that could be None could make it easier to use.

**Comment by rb-determined-ai** [comment]
> Up to you

I chose to remove the footgun, using the `"attr" in vars(self)` strategy.  Footguns are just bad, even if it takes an antipattern to trigger them.

> having the option of passing a value that could be None could make it easier to use.

I think I'm confused about what you are saying though.  There's only two behaviors:
* `omit_empty=False`: emit the explicitly set value or the implicit default (which happens to always be None)
* `omit_empty=True`: emit the explicitly set value, or nothing if there is no explicitly set value

So what would `to_json(omit_emtpy=None)` even mean?

**Comment by hamidzr** [comment]
> > having the option of passing a value that could be None could make it easier to use.
> 
> I think I'm confused about what you are saying though. There's only two behaviors:
> 
> * `omit_empty=False`: emit the explicitly set value or the implicit default (which happens to always be None)
> * `omit_empty=True`: emit the explicitly set value, or nothing if there is no explicitly set value
> 
> So what would `to_json(omit_emtpy=None)` mean?

image

depending on the use case I could this be helpful or a footgun? if you don't notice you could be passing in a None

**Review by hamidzr — APPROVED** [review]

**Comment by rb-determined-ai** [comment]
@hamidzr but a signature like:

```python
def foo(a: Optional[bool] = None) -> None:
```
suggests 3 possible behaviors, while:

```python
def foo(a: bool = False)
```

suggests 2 behaviors.

There's only two behaviors in this case, so the typing should reflect that.  Mypy _should_ raise an error if you pass `None`, because it doesn't make sense.  By extension, passing `Optional[bool]` also doesn't make sense.  Sure, it will act like `False` at runtime, but it should not pass type checks.

**Comment by hamidzr** [comment]
that makes sense. I brought it up mostly because I was curious how I should look at default args in py. By doing `=False` it feels as if you're setting None as an acceptable input based on how I usually see kwargs set up in python but no question that just `bool` is more accurate.
but maybe this was more relevant when python didn't have explicit type hints. eg 
```python
fn(a=3):

fn(a=None):
  a = 3 if a is None..
```
